evolver/device.py

```diff
@@ -70,26 +74,49 @@ def schema(self):
             "controllers": [{"kind": str(type(a)), "config": a.Config.model_json_schema()} for a in self.controllers],
         }
 
+    def _loop_exception_wrapper(self, callable, message="unknown") -> bool:
+        try:
+            callable()
+            return False
+        except Exception:
+            self.logger.exception(f"Error in loop: {message}:")
+            if self.raise_loop_exceptions:
+                raise
+            return True
+
     def read_state(self):
+        read_error = False
         for name, device in self.sensors.items():
-            device.read()
+            err = self._loop_exception_wrapper(device.read, f"reading device {name}")
+            read_error = read_error or err
             self.last_read[name] = time.time()
             self.history.put(name, device.get())
+        return read_error
 
     def evaluate_controllers(self):
-        for controller in self.controllers:
-            controller.run()
+        errs = [self._loop_exception_wrapper(c.run, f"updating controller {c}") for c in self.controllers]
+        return any(errs)
 
     def commit_proposals(self):
-        for device in self.effectors.values():
-            device.commit()
+        errs = [
+            self._loop_exception_wrapper(d.commit, f"committing proposals for {d}") for d in self.effectors.values()
+        ]
+        return any(errs)
 
     def loop_once(self):
-        self.read_state()
-        # for any hardware awaiting calibration, call calibration update method here
+        read_error = self.read_state()
+        if read_error and self.skip_control_on_read_failure:
+            self.logger.info("Skipping control loop due to read error")
+            return
         if self.enable_control:
-            self.evaluate_controllers()
-            self.commit_proposals()
+            control_error = self.evaluate_controllers()
+            if control_error and self.abort_on_control_errors:
+                self.abort()
+                raise RuntimeError("Aborted due to control error")
```

Comment:
This raises _an_ error but not _the_ error(s) and it would be cool if we could chain them. If the funcs returned a list of errors rather than just a bool, we could use ``from``, e.g., ``raise RuntimeError("Aborted due to control error") from control_error[0]`` (or chain them all - though the 1st might be enough, with a note that there could be more).

Reply:
I think runtime error would be fine. I can do the exception from thing, I'll keep it simple and just raise from first one and say see logs.